### PR TITLE
Linear regression window operation

### PR DESCRIPTION
**Pull request by mtopolnik**
Also improve mutable `Accumulators`.

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import java.io.IOException;
+import java.math.BigInteger;
+
+/**
+ * Maintains the components needed to compute the linear regression on a
+ * set of {@code (long, long)} pairs. The intermediate results are held in
+ * the form of {@link BigInteger} and the finished value is a {@code
+ * double}-valued linear coefficient.
+ */
+public final class LinRegAccumulator {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: There are two components to linear regression: slope and y-position: `y = ax + b`, `a` being the slope. We should call this `LinearRegressionSlope` or calculate both.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        sumXY = sumXY.subtract(that.sumXY);
+        sumX2 = sumX2.subtract(that.sumX2);
+        return this;
+    }
+
+    /**
+     * Computes the linear coefficient of the linear regression of the
+     * accumulated samples.
+     */
+    public double finish() {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: We will have division by zero, if N<=1.
If all pairs had the same `x`-value (this is when the denominator is zero), the line is vertical and has infinite slope.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ * Maintains the components needed to compute the linear regression on a
+ * set of {@code (long, long)} pairs. The intermediate results are held in
+ * the form of {@link BigInteger} and the finished value is a {@code
+ * double}-valued linear coefficient.
+ */
+public final class LinRegAccumulator {
+    private static final int MAX_BIGINT_LEN = 255;
+
+    private long n;
+    private BigInteger sumX;
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: We could use int128, it's much faster than BigInteger. Normally, X will be time. We add `X^2`, n-times. If `x` is `long`, then we need int192. I experimented with this one: https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/SignedInt128.java, when I was doing my original streaming sample, it was considerably faster.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java (hunk 2 above)
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: Actually,  this is correct. We'll never have n=0, as we don't emit windows, when there were no events.
With n=1, we'll have `0 / 0`, which is `NaN`, which is correct. With `n>1` and all x-values the same, we'll have `a / 0`, which is `Inf` or `-Inf`, which is also correct to emit. Anyway, I'll add this to the test.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java (hunk 1 above)
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: I'm fine with the current name... if we ever need to differentiate, we'll rename to more specific names.

**Comment by cangencer** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java (hunk 3 above)
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: let's defer this level of optimization for some other time

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    private BigInteger sumY;
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/accumulator/LinRegAccumulator.java]: We shouldn't since the sum is always an integer.

**Pull request merged**